### PR TITLE
Desktop session recording/playback

### PR DESCRIPTION
**Pull request by zmb3**
Here we introduce a new protobuf type (DesktopRecording) that contains
an encoded TDP message, and update AuditWriter to treat these similarly
to SessionPrint events (which are used for SSH session recordings).

Additionally, a new API server endpoint streams TDP messages from these events
over a websocket interface where they can be rendered by the browser.

Updates #9978
Updates #9979

### CODE REVIEW

File: lib/events/auditwriter.go [hunk 1]
```diff
@@ -139,6 +145,32 @@ func (cfg *AuditWriterConfig) CheckAndSetDefaults() error {
 	if cfg.BackoffDuration == 0 {
 		cfg.BackoffDuration = defaults.NetworkBackoffDuration
 	}
+	if cfg.MakeEvents == nil {
```
Comment (zmb3) [inline lib/events/auditwriter.go:148]: Note: we technically don't need this, as it's only used when you use `AuditWriter`'s `io.Writer` interface, and we are calling `EmitAuditEvent()` directly.

However, it felt like a trap to have a general purpose tool called `AuditWriter` that had special case logic for SSH sessions and didn't provide a way to handle desktop or other types of sessions.

I could be convinced to remove this. Longer term, we should probably have a separate type that wraps `AuditWriter` and provides the `io.Writer` interface where necessary.

File: lib/srv/desktop/tdp/conn.go [hunk 2]
```diff
@@ -29,6 +29,11 @@ import (
 type Conn struct {
 	rw   io.ReadWriter
 	bufr *bufio.Reader
+
+	// OnSend is an optional callback that is invoked when a TDP message
+	// is sent on the wire. It is passed both the raw bytes and the encoded
```
Comment (ibeckermayer) [inline lib/srv/desktop/tdp/conn.go:34]:
```suggestion
	// is sent on the wire (when OutputMessage is called). It is passed both the raw bytes and the encoded
```

File: lib/events/auditwriter.go (hunk 1 above)
Comment (timothyb89) [reply lib/events/auditwriter.go:148]: I don't especially mind this but it might be nice to keep the default behavior in a named function rather than embedded in a closure here, needing to keep a field `nil` to select a particular implementation seems like an unusual API convention.

**Review by timothyb89 — APPROVED** [review]

**Review by lxea — APPROVED** [review]

File: lib/srv/desktop/windows_server.go [hunk 3]
```diff
@@ -333,6 +340,7 @@ func NewWindowsService(cfg WindowsServiceConfig) (*WindowsService, error) {
 		Clock: s.cfg.Clock,
 	})
 	if err != nil {
+		s.Close()
```
Comment (codingllama) [inline lib/srv/desktop/windows_server.go:343]: This might be a good place to apply that "!ok/close" idiom:

```go
// Declare closeable resources
ok := false
defer func() {
  if !ok {
    // Close resources
  }
}()

// Method continues ...

ok = true
return
```

File: lib/srv/desktop/windows_server.go [hunk 4]
```diff
@@ -644,6 +709,33 @@ func (s *WindowsService) connectRDP(ctx context.Context, log logrus.FieldLogger,
...
+		log.Infof("desktop session %v will not be recorded, user %v's roles disable recording", string(sessionID), authCtx.User.GetName())
+	}
+
+	sw, err := s.newStreamWriter(recordSession, string(sessionID))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	defer func() {
+		go func() {
```
Comment (codingllama) [inline lib/srv/desktop/windows_server.go:732]: Maybe leave a comment here? A `go` inside a `defer` is not that common.

File: lib/srv/desktop/windows_server.go [hunk 5]
```diff
@@ -660,7 +752,45 @@ func (s *WindowsService) connectRDP(ctx context.Context, log logrus.FieldLogger,
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
+	delay := timer()
 	tdpConn := tdp.NewConn(proxyConn)
+	sessionStartTime := s.cfg.Clock.Now().UTC().Round(time.Millisecond)
+	tdpConn.OnSend = func(m tdp.Message, b []byte) {
```
Comment (codingllama) [inline lib/srv/desktop/windows_server.go:758]: Maybe give the event handlers names so there is less noise inside connectRDP? That way you could also test them separately.

File: lib/srv/desktop/windows_server.go [hunk 6]
```diff
@@ -660,7 +752,45 @@ func (s *WindowsService) connectRDP(ctx context.Context, log logrus.FieldLogger,
...
+				Message:           b,
+				DelayMilliseconds: delay(),
+			}); err != nil {
+				s.cfg.Log.WithError(err).Warning("could not emit desktop recording event")
+			}
+		}
+	}
+	tdpConn.OnRecv = func(m tdp.Message) {
+		switch m.(type) {
+		// TODO: tdp.ClipboardData
```
Comment (codingllama) [inline lib/srv/desktop/windows_server.go:775]:
```suggestion
		// TODO(zmb3): Record audit receive events for tdp.ClipboardData messages
```

File: lib/srv/desktop/windows_server.go [hunk 7]
```diff
@@ -660,7 +752,45 @@ func (s *WindowsService) connectRDP(ctx context.Context, log logrus.FieldLogger,
...
+		case tdp.ClientScreenSpec, tdp.MouseButton, tdp.MouseMove:
+			b, err := m.Encode()
+			if err != nil {
+				s.cfg.Log.WithError(err).Warning("could not emit desktop recording event")
```
Comment (codingllama) [inline lib/srv/desktop/windows_server.go:779]: Should we record the fact that the encode operation failed? It is worth having that and/or the error in the event?

File: lib/web/desktop_playback.go [hunk 8]
```diff
@@ -0,0 +1,246 @@
+/*
+Copyright 2021 Gravitational, Inc.
```
Comment (codingllama) [inline lib/web/desktop_playback.go:2]:
```suggestion
Copyright 2022 Gravitational, Inc.
```

I would only bother with new files, though.

File: lib/web/desktop_playback.go [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+	"github.com/sirupsen/logrus"
+	"golang.org/x/net/websocket"
+)
+
+const actionPlayPause = "play/pause"
+
+const (
+	playStatePlaying  = "playing"
+	playStatePaused   = "paused"
+	playStateFinished = "finished"
```
Comment (codingllama) [inline lib/web/desktop_playback.go:43]: Do we want to strongly type these? (Instead of string.)

File: lib/web/desktop_playback.go [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+)
+
+// playbackAction is a message passed from the playback client
+// to the server over the websocket connection in order to modify
+// the playback state.
+type playbackAction struct {
+	// Action is one of actionPlayPause | TODO: actionMove
```
Comment (codingllama) [inline lib/web/desktop_playback.go:50]: This comment looks a bit off. Move the TODO to another line and the part about actionPlayPause to the const itself?

File: lib/web/desktop_playback.go [hunk 11]
```diff
@@ -0,0 +1,246 @@
+/*
```
Comment (codingllama) [inline lib/web/desktop_playback.go:1]: Do we have a better package for this file than lib/web? I'm trying to avoid adding to the more "general" packages.

File: lib/web/desktop_playback.go [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+// to the server over the websocket connection in order to modify
+// the playback state.
+type playbackAction struct {
+	// Action is one of actionPlayPause | TODO: actionMove
+	// actionPlayPause toggles the playbackState.playState
+	Action string `json:"action"`
+}
+
+// playbackPlayer is a thread-safe struct for managing
+// the global state of the playback websocket connection.
```
Comment (codingllama) [inline lib/web/desktop_playback.go:56]: What is the "playback websocket connection"? Could you give it more context?

File: lib/web/desktop_playback.go [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+type playbackPlayer struct {
```
Comment (codingllama) [inline lib/web/desktop_playback.go:57]: Does this plug into anything? Should it be in another PR?

Could we have tests too?  ;)

File: lib/web/desktop_playback.go [hunk 14]
```diff
@@ -0,0 +1,246 @@
...
+				pp.log.Warnf("session %v contains unexpected event type %T", pp.sID, evt)
+			}
+		}
+	}
+}
+
+// Play kicks off goroutines for receiving actions
+// and playing back the session over the websocket,
+// and then hangs until the websocket is closed.
+func (pp *playbackPlayer) Play(ctx context.Context) {
```
Comment (codingllama) [inline lib/web/desktop_playback.go:216]: Move this higher to the top? The rest made a lot more sense once I read Play.

File: lib/web/desktop_playback.go [hunk 15]
```diff
@@ -0,0 +1,246 @@
...
+	defer pp.log.Debug("playbackPlayer.Play returned")
+	ppCtx, cancel := context.WithCancel(ctx)
+	defer pp.close(cancel)
```
Comment (codingllama) [inline lib/web/desktop_playback.go:219]: I found passing an external cancel like this a bit strange, specially when reading the methods that take CancelFunc as an argument.

If we are not allowing players to be reused (which I think is the case), then maybe consider having an internal cancel that the funcs can access?

File: lib/events/auditwriter.go (hunk 1 above)
Comment (codingllama) [reply lib/events/auditwriter.go:148]: Seems reasonable to me too, although I'm not intimately familiar with these parts. A named func would be nice (and testable!)

**Review by codingllama — COMMENTED** [review]
Sorry about the delay, reviewed.

File: api/types/events/events.proto [hunk 16]
```diff
@@ -230,6 +230,20 @@ message SessionPrint {
...
+message DesktopRecording {
+    // Metadata is a common event metadata
+    Metadata Metadata = 1
+        [ (gogoproto.nullable) = false, (gogoproto.embed) = true, (gogoproto.jsontag) = "" ];
+
+    // Message is the encoded TDP message.
+    bytes Message = 2 [ (gogoproto.nullable) = true, (gogoproto.jsontag) = "message" ];
+
+    // DelayMilliseconds is the delay in milliseconds from the start of the session
+    int64 DelayMilliseconds = 3 [ (gogoproto.jsontag) = "ms" ];
```
[the remark this answers is not in the document]
Comment (zmb3) [reply api/types/events/events.proto:244]: I agree. In this case we were trying to keep compatibility with `message SessionPrint` (which is the equivalent protobuf for SSH sessions).

Given that context, are you okay with keeping this or still want to see it changed?

In all current use cases, the front end is either expecting SSH or desktop, so they don't necessarily need to match, but was being cautious as we didn't know how the front end would play out when we first added recording.

File: api/types/events/events.proto [hunk 17]
```diff
@@ -1676,6 +1690,24 @@ message WindowsDesktopSessionEnd {
     string WindowsUser = 7 [ (gogoproto.jsontag) = "windows_user" ];
     // DesktopLabels are the labels on the desktop resource.
     map<string, string> DesktopLabels = 8 [ (gogoproto.jsontag) = "desktop_labels" ];
+    // StartTime is the timestamp at which the session began.
+    google.protobuf.Timestamp StartTime = 9 [
+        (gogoproto.stdtime) = true,
+        (gogoproto.nullable) = false,
+        (gogoproto.jsontag) = "session_start,omitempty"
```
[the remark this answers is not in the document]
Comment (zmb3) [reply api/types/events/events.proto:1697]: Same here, following the unfortunate precedent set by `message SessionEnd`.

File: lib/events/api.go [hunk 18]
```diff
@@ -453,9 +453,12 @@ const (
 	// WindowsDesktopSessionStartEvent is emitted when a user attempts
 	// to connect to a desktop.
 	WindowsDesktopSessionStartEvent = "windows.desktop.session.start"
-	// WindowsDesktopSessionEndEvent is emitted when a user  disconnects
+	// WindowsDesktopSessionEndEvent is emitted when a user disconnects
 	// from a desktop.
 	WindowsDesktopSessionEndEvent = "windows.desktop.session.end"
+
```
[the remark this answers is not in the document]
Comment (zmb3) [reply lib/events/api.go:459]: Eh.`DesktopRecordingEvent` is OS-agnostic, and would work if we end up supporting desktops on macOS or Linux. This is why it was not grouped with the Windows-specific events.

[the remark this answers is not in the document]
Comment (codingllama) [reply lib/events/api.go:459]: Makes sense, my bad.

File: api/types/events/events.proto (hunk 16 above)
[the remark this answers is not in the document]
Comment (codingllama) [reply api/types/events/events.proto:244]: Right, I thought it might be something like this. Up to you, but maybe leave a comment so context is not lost?

File: lib/srv/desktop/windows_server.go (hunk 3 above)
Comment (zmb3) [reply lib/srv/desktop/windows_server.go:343]: I like it!

File: lib/srv/desktop/windows_server.go (hunk 7 above)
Comment (zmb3) [reply lib/srv/desktop/windows_server.go:779]: As of now, this can basically never happen, as encode just writes to a `*bytes.Buffer`, which is guaranteed to always return a `nil` error.

File: lib/web/desktop_playback.go (hunk 11 above)
Comment (codingllama) [reply lib/web/desktop_playback.go:1]: I like that too, sounds good.

File: lib/web/desktop_playback.go (hunk 15 above)
Comment (zmb3) [reply lib/web/desktop_playback.go:219]: This came about because I'm not a fan of holding on to `context.Context` (or its associated cancel funcs) as struct members and asked Isaiah to see if we could pass them as arguments instead.

In this case, I guess we could capture the CancelFunc as a member of the player struct, but my newfound Rust experience is telling me this is a bad idea. The context is only valid for the duration of `Play`, so we shouldn't stash the associated cancel somewhere that could outlive `Play`.

Have any ideas?

File: api/types/events/events.proto (hunk 16 above)
[the remark this answers is not in the document]
Comment (zmb3) [reply api/types/events/events.proto:244]: fair, will do

File: lib/web/desktop_playback.go (hunk 15 above)
Comment (codingllama) [reply lib/web/desktop_playback.go]: If we don't allow more than a single invocation of Play, which seems to be the intent, then my initial idea would be to nest playCtx/playCancel fields inside playbackPlayer, initialized on new.

In general I don't like concurrency primitives passed around - funcs with channels, locks or cancels as params. I'd either handle concurrency where the goroutines are launched (ie, sync primitives live in a single method) or nest attributes. That said, what you have definitely works and it's all private, so feel free to push back if you don't think this is worth pursuing.

File: lib/web/desktop/playback.go [hunk 19]
```diff
@@ -0,0 +1,240 @@
...
+	"github.com/gravitational/teleport/lib/session"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/net/websocket"
+)
+
+// Player manages the playback of a recorded desktop session.
+// It streams events from the audit log to the browser over
+// a websocket connection.
+type Player struct {
```
Comment (codingllama) [inline lib/web/desktop/playback.go:38]: Oh yes. Separate package, public methods make a coherent interface, looks great. :D

**Review by codingllama — COMMENTED** [review]
Give me a nudge when you are done with the comments.

**Review by zmb3 — COMMENTED** [review]
@codingllama mind taking one more pass?

I added a simple streaming test. We'll be moving from `x/net/websocket` to `gorilla/websocket` after this PR, at which time I'll be able to test more sophisticated error and cancelation scenarios.

File: lib/web/desktop_playback.go (hunk 15 above)
Comment (zmb3) [reply lib/web/desktop_playback.go]: I don't love either approach, to be honest. Right now, we don't allow more than one invocation of `Play()`. But that might change. 

I'd prefer that we continue to pass the context to play, as that is the long running operation that must support cancelation, and I can't really think of a better way to maintain that invariant without passing around a shared cancelation signal.

**Review by codingllama — APPROVED** [review]

File: lib/srv/desktop/windows_server_test.go [hunk 20]
```diff
@@ -192,3 +195,83 @@ func TestGenerateCredentials(t *testing.T) {
...
+		var msg tdp.Message = nil
+		handler(msg, encoded)
+	}()
+
+	select {
+	case e := <-ch:
+		dr, ok := e.(*events.DesktopRecording)
+		require.True(t, ok)
+		require.Equal(t, encoded, dr.Message)
+	case <-time.After(1 * time.Second):
```
[the remark this answers is not in the document]
Comment (zmb3) [reply lib/srv/desktop/windows_server_test.go:226]: 1 second is an eternity for a channel send. I was aiming for a quick failure here rather than a failure that causes CI to to hang until the max timeout. 🤷‍♂️

[the remark this answers is not in the document]
Comment (codingllama) [reply lib/srv/desktop/windows_server_test.go:226]: Fair.

**Review by codingllama — APPROVED** [review]
